### PR TITLE
socket, accept: set CLOEXEC on platforms that support it

### LINKED ISSUE #16180: create sockets and open files with CLOEXEC by default
Sockets should not be shared with child processes. So probably it makes sense to add SOCK_CLOEXEC when creating a socket.

https://github.com/dotnet/corefx/blob/7a230270559cfbef056aaafc022cdfd8fd3d829e/src/Native/Unix/System.Native/pal_networking.cpp#L2363

### PR DESCRIPTION
**Pull request by tmds**
Fixes #16180

### CODE REVIEW

**Comment by tmds** [comment]
@stephentoub @Priya91 PTAL

- No one is using Sys.OpenPipe, should I remove it? There is a Sys.Pipe which is used.
- Q unrelated to this PR: most functions that can return EINTR are wrapped with CheckInterrupted. close in SystemNative_Close isn't. Is this intentional? Or an oversight?

File: src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs [hunk 1]
```diff
@@ -45,7 +45,7 @@ private SafeFileHandle OpenHandle(FileMode mode, FileShare share, FileOptions op
                 _asyncState = new AsyncState();
 
             // Translate the arguments into arguments for an open call.
-            Interop.Sys.OpenFlags openFlags = PreOpenConfigurationFromOptions(mode, _access, options); // FileShare currently ignored
+            Interop.Sys.OpenFlags openFlags = PreOpenConfigurationFromOptions(mode, _access, share, options); // FileShare currently ignored
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: @JeremyKuhne, @jkotas, is the FileStream code in corefx being used by anyone at this point?  Can we delete it all from corefx?

File: src/Native/Unix/System.Native/pal_console.cpp [hunk 2]
```diff
@@ -430,7 +431,11 @@ static bool InitializeSignalHandling()
     // thread.  We can't do anything interesting in the signal handler,
     // so we instead send a message to another thread that'll do
     // the handling work.
+#if HAVE_PIPE2
+    if (pipe2(g_signalPipe, O_CLOEXEC) != 0)
+#else
     if (pipe(g_signalPipe) != 0)
+#endif
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_console.cpp]: Can you just use SystemNative_Pipe instead?  It already uses pipe2 if available or else uses pipe and then fcntl with FD_CLOEXEC.

File: src/Native/Unix/System.Native/pal_networking.cpp [hunk 3]
```diff
@@ -1730,7 +1730,11 @@ extern "C" Error SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, in
 
     socklen_t addrLen = static_cast<socklen_t>(*socketAddressLen);
     int accepted;
+#if defined(HAVE_ACCEPT_4) && defined(SOCK_CLOEXEC)
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.cpp]: Do any of the systems we currently target not have these?  Do any of those currently being brought up by the community (e.g. Android) not have these?

File: src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: @tmds, the FileStream changes will need to be made in the coreclr repo for them to affect .NET Core.

**Comment by stephentoub** [comment]
> No one is using Sys.OpenPipe, should I remove it? There is a Sys.Pipe which is used.

Yes, please. Thanks.

> most functions that can return EINTR are wrapped with CheckInterrupted. close in SystemNative_Close isn't. Is this intentional? Or an oversight?

It's intentional.  From the man page:
```
       The EINTR error is a somewhat special case.  Regarding the EINTR
       error, POSIX.1-2013 says:

              If close() is interrupted by a signal that is to be caught, it
              shall return -1 with errno set to EINTR and the state of
              fildes is unspecified.

       This permits the behavior that occurs on Linux and many other
       implementations, where, as with other errors that may be reported by
       close(), the file descriptor is guaranteed to be closed.  However, it
       also permits another possibility: that the implementation returns an
       EINTR error and keeps the file descriptor open.  (According to its
       documentation, HP-UX's close() does this.)  The caller must then once
       more use close() to close the file descriptor, to avoid file
       descriptor leaks.  This divergence in implementation behaviors
       provides a difficult hurdle for portable applications, since on many
       implementations, close() must not be called again after an EINTR
       error, and on at least one, close() must be called again.  There are
       plans to address this conundrum for the next major release of the
       POSIX.1 standard.
```

File: src/Native/Unix/System.Native/pal_networking.cpp (hunk 3 above)
Comment (tmds) [inline src/Native/Unix/System.Native/pal_networking.cpp]: I expect this to be present on Linux systems where .NET Core is used. It has been around for a while.
Mac doesn't have it.

File: src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs (hunk 1 above)
Comment (jkotas) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: > Can we delete it all from corefx?

I think so. https://github.com/dotnet/corefx/pull/17229

Comment (tmds) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: coreclr PR: https://github.com/dotnet/coreclr/pull/10251

File: src/Native/Unix/System.Native/pal_io.h [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include "pal_types.h"
+#include "pal_errno.h"
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.h]: What is this needed for?

File: src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/FileStream.Unix.cs]: @tmds, can you remove the changes to these files from this PR?  It'll conflict with Jan's change to delete the files entirely.

File: src/Native/Unix/System.Native/pal_io.h (hunk 4 above)
Comment (tmds) [inline src/Native/Unix/System.Native/pal_io.h]: pal_io.h uses the Error enum defined in pal_errno.h

Comment (tmds) [inline src/Native/Unix/System.Native/pal_io.h]: ... pal_console.cpp now includes pal_io.h for the SystemNative_Pipe, but it doesn't include pal_errno.h so it gave a compile error.
The other cpp files including pal_io.h must have included pal_errno.h before it.

Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.h]: Ah,ok

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop OSX10.12 Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please

**Comment by Priya91** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test

[OSX failure has timed out with hudson error.]

**Comment by tmds** [comment]
Thanks @stephentoub and @Priya91

**Pull request merged**